ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PullQueryRoutingFunctionalTest.java

```diff
@@ -480,5 +553,94 @@ private String extractQueryId(final String outputString) {
     assertThat("Could not find query id in: " + outputString, matcher.find());
     return matcher.group(1);
   }
+
+  public static class TestApp {
+
+    private final KsqlHostInfoEntity host;
+    private final TestKsqlRestApp app;
+    private final Shutoffs shutoffs;
+
+    public TestApp(KsqlHostInfoEntity host, TestKsqlRestApp app, Shutoffs shutoffs) {
+      this.host = host;
+      this.app = app;
+      this.shutoffs = shutoffs;
+    }
+
+    public KsqlHostInfoEntity getHost() {
+      return host;
+    }
+
+    public TestKsqlRestApp getApp() {
+      return app;
+    }
+
+    public Shutoffs getShutoffs() {
+      return shutoffs;
+    }
+  }
+
+  public static class Shutoffs {
+    private final AtomicBoolean ksqlOutgoing = new AtomicBoolean(false);
+    private final AtomicBoolean kafkaIncoming = new AtomicBoolean(false);
+
+    public void shutOffAll() {
```

Comment:
This stops all incoming/outgoing traffic to simulate that a node is dead? Does it actually cause rebalancing?

Reply:
No, it doesn't technically stop all traffic to kafka.  I couldn't figure out a very simple way to do that.  My understanding given https://www.confluent.io/blog/cooperative-rebalancing-in-kafka-streams-consumer-ksqldb/ is that `JoinGroup` is another request type used to coordinate groups, which is different from consumption.

While this test isn't 100% true to a production environment, I also didn't want a pause in the test to result in rebalancing and weird, unpredictable results.  Also, it's hard to both kill off a server, and know that it will take effect immediately.  You can call stop, which will potentially block a bit or you can run it in a background thread, in which case you don't know when it will take effect.

For that reason, I chose to simulate things in this way, but tell me if you disagree.

Reply:
Yes, that makes sense. 

I was trying to see if we can test everything using these functional tests and avoid creating cloud tests. But it seems, that if we cannot test the actual rebalancing and measure for example how long it takes for the heartbeat to find a host is down, how long it takes for pull queries to succeed at standby, find that routing correctly works after rebalancing completes (first go to standby, then go to new active) etc, we may still need to perform cloud tests.